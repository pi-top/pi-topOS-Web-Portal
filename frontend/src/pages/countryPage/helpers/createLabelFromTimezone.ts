export default (timezone: string) => {
  if (!timezone) {
    return '';
  }

  return timezone
    .replace(/_/g, ' ')
    .replace(/\//, ' - ')
    .replace(/\//g, ', ');
};
