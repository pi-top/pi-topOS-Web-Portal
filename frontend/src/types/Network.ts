export type Network = {
  ssid: string;
  passwordRequired: boolean
};

export type NetworkCredentials = {
  ssid: string;
  password: string
};
