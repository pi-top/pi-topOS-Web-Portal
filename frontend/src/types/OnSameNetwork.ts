export type OnSameNetworkData = {
  onSameNetwork: boolean;
  piTopIp: string;
  clientIp: string;
}
