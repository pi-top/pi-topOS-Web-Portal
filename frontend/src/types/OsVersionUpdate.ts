export type OsVersionUpdate = {
  shouldBurn: boolean;
  requireBurn: boolean;
  latestOSVersion: string;
  update: boolean;
};
