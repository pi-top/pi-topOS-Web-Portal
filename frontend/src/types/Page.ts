export enum Page {
  Splash = "SPLASH",
  Wifi = "WIFI",
  Upgrade = "UPGRADE",
  Country = "COUNTRY",
  Keyboard = "KEYBOARD",
  Language = "LANGUAGE",
  Registration = "REGISTRATION",
  Restart = "RESTART",
  Terms = "TERMS",
  LandingSplash = "LANDINGSPLASH",
  Links = "LINKS",
}

export enum PageRoute {
  Splash = "/onboarding",
  Wifi = "/onboarding/wifi",
  Upgrade = "/onboarding/upgrade",
  Country = "/onboarding/country",
  Keyboard = "/onboarding/keyboard",
  Language = "/onboarding/language",
  Registration = "/onboarding/registration",
  Restart = "/onboarding/reboot",
  Terms = "/onboarding/terms",
  Links = "/landing/links",
  LandingSplash = "/landing"
}
