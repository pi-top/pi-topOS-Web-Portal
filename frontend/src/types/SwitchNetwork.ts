export type SwitchNetworkData = {
  shouldSwitchNetwork: boolean;
  shouldDisplayDialog: boolean;
  piTopIp: string;
  clientIp: string;
}
