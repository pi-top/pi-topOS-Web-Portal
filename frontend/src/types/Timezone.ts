export type Timezone = {
  countryCode: string;
  timezone: string;
};
