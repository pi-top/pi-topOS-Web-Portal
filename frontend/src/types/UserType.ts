export enum UserType {
  Home = "HOME",
  School = "SCHOOL",
}
