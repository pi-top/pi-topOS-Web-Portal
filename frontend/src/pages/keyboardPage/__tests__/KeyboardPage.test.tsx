import React from "react";
import {
  render,
  fireEvent,
  BoundFunction,
  GetByBoundAttribute,
  QueryByText,
  GetByText,
  RenderResult,
  QueryByBoundAttribute,
} from "@testing-library/react";

import KeyboardPage, { Props, ErrorMessage } from "../KeyboardPage";
import openReactSelectMenu from "../../../../test/helpers/openReactSelectMenu";
import keyboards from "./data/keyboards.json";
import keyboardVariants from "./data/keyboardVariants.json";
import { Keyboard } from "../../../types/Keyboard";
import querySpinner from "../../../../test/helpers/querySpinner";
import { createLabelFromKeyboard } from "../helpers/keyboardSelect";
import queryReactSelect from "../../../../test/helpers/queryReactSelect";
import reactSelectIsDisabled from "../../../../test/helpers/reactSelectIsDisabled";

describe("KeyboardPage", () => {
  let defaultProps: Props;
  let keyboardPage: HTMLElement;
  let getByAltText: BoundFunction<GetByBoundAttribute>;
  let queryByText: BoundFunction<QueryByText>;
  let getByText: BoundFunction<GetByText>;
  let queryByTestId: BoundFunction<QueryByBoundAttribute>;
  let rerender: RenderResult["rerender"];
  beforeEach(() => {
    defaultProps = {
      keyboards: {},
      keyboardVariants: {},
      currentKeyboard: {
        layout: "us",
      },
      isGettingKeyboards: false,
      setKeyboard: jest.fn(),
      settingKeyboard: null,
      getKeyboardsError: false,
      setKeyboardError: false,
      onBackClick: jest.fn(),
      onSkipClick: jest.fn(),
      onNextClick: jest.fn(),
      alwaysAllowSkip: false,
    };

    ({
      container: keyboardPage,
      getByAltText,
      queryByText,
      queryByTestId,
      getByText,
      rerender,
    } = render(<KeyboardPage {...defaultProps} />));
  });

  it("renders correct image", () => {
    expect(getByAltText("keyboard-screen")).toMatchSnapshot();
  });

  it("renders prompt correctly", () => {
    const prompt = keyboardPage.querySelector(".prompt");
    expect(prompt).toMatchSnapshot();
  });

  it("renders explanation", () => {
    expect(
      queryByText("Which keyboard layout would you like to use?")
    ).toBeInTheDocument();
  });

  it("does not render error message", () => {
    expect(queryByTestId("error-message")).not.toBeInTheDocument();
  });

  it("renders next button", () => {
    expect(queryByText("Next")).toBeInTheDocument();
  });

  it("does not render skip button", () => {
    expect(queryByText("Skip")).not.toBeInTheDocument();
  });

  it("renders current keyboard as selected option", () => {
    expect(
      queryByText(
        createLabelFromKeyboard(
          defaultProps.currentKeyboard,
          keyboards,
          keyboardVariants
        )
      )
    ).toBeInTheDocument();
  });

  describe("when isGettingKeyboards is true", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        isGettingKeyboards: true,
      };

      rerender(<KeyboardPage {...defaultProps} />);
    });

    it("does not render select", () => {
      expect(queryReactSelect(keyboardPage)).not.toBeInTheDocument();
    });

    it("disables the next button", () => {
      expect(getByText("Next").parentElement).toBeDisabled();
    });
  });

  describe("when settingKeyboard is set", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        settingKeyboard: {
          layout: "gb",
        },
      };

      rerender(<KeyboardPage {...defaultProps} />);
    });

    it.only("disables next button", () => {
      expect(queryByText("Next")?.parentElement).toBeDisabled();
    });

    it("sets settingKeyboard as selected option", () => {
      expect(
        queryByText(
          createLabelFromKeyboard(
            defaultProps.settingKeyboard!,
            keyboards,
            keyboardVariants
          )
        )
      ).toBeInTheDocument();
    });

    it('disables select', () => {
      expect(reactSelectIsDisabled(keyboardPage)).toEqual(true);
    })

    it("renders setting keyboard message", () => {
      expect(queryByText("setting keyboard...")).toBeInTheDocument();
    });

    it("renders loading spinner", () => {
      expect(querySpinner(keyboardPage)).toBeInTheDocument();
    });
  });

  describe("when getKeyboardsError is true", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        getKeyboardsError: true,
      };

      rerender(<KeyboardPage {...defaultProps} />);
    });

    it("renders correct error message", () => {
      expect(queryByText(ErrorMessage.GetKeyboards)).toBeInTheDocument();
    });
  });

  describe("when setKeyboardError is true", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        setKeyboardError: true,
      };

      rerender(<KeyboardPage {...defaultProps} />);
    });

    it("renders correct error message", () => {
      expect(queryByText(ErrorMessage.SetKeyboard)).toBeInTheDocument();
    });

    it("renders skip button", () => {
      expect(queryByText("Skip")).toBeInTheDocument();
    });

    it("calls onSkipClick when skip button clicked", () => {
      fireEvent.click(getByText("Skip"));

      expect(defaultProps.onSkipClick).toHaveBeenCalled();
    });
  });

  describe("when alwaysAllowSkip is true", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        alwaysAllowSkip: true,
      };

      rerender(<KeyboardPage {...defaultProps} />);
    });

    it("renders skip button", () => {
      expect(queryByText("Skip")).toBeInTheDocument();
    });

    it("calls onSkipClick when skip button clicked", () => {
      fireEvent.click(getByText("Skip"));

      expect(defaultProps.onSkipClick).toHaveBeenCalled();
    });
  });

  describe("when keyboards are passed", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        keyboards: { ...keyboards },
      };

      rerender(<KeyboardPage {...defaultProps} />);
    });

    it("renders correct keyboard select options", async () => {
      openReactSelectMenu(
        keyboardPage,
        createLabelFromKeyboard(
          defaultProps.currentKeyboard,
          keyboards,
          keyboardVariants
        )
      );

      Object.keys(defaultProps.keyboards).forEach((layout) => {
        // exclude current keyboard as it is already in the select input
        if (layout === defaultProps.currentKeyboard.layout) return;

        expect(
          queryByText(
            createLabelFromKeyboard(
              {
                layout,
              },
              keyboards,
              keyboardVariants
            )
          )
        ).toBeInTheDocument();
      });
    });
  });

  describe("when keyboards and keyboardVariants are passed", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        keyboards: { ...keyboards },
        keyboardVariants: { ...keyboardVariants },
      };

      rerender(<KeyboardPage {...defaultProps} />);
    });

    it("renders correct keyboard select options", async () => {
      openReactSelectMenu(
        keyboardPage,
        createLabelFromKeyboard(
          defaultProps.currentKeyboard,
          keyboards,
          keyboardVariants
        )
      );

      Object.keys(defaultProps.keyboards).forEach((layout) => {
        // exclude current keyboard as it is already in the select input
        if (layout === defaultProps.currentKeyboard.layout) return;

        expect(
          queryByText(
            createLabelFromKeyboard(
              {
                layout,
              },
              keyboards,
              keyboardVariants
            )
          )
        ).toBeInTheDocument();

        // @ts-ignore
        const variants = keyboardVariants[layout];
        if (variants) {
          expect(
            queryByText(
              createLabelFromKeyboard(
                {
                  layout,
                  // @ts-ignore
                  variant: Object.keys(variants)[0],
                },
                keyboards,
                keyboardVariants
              )
            )
          );
        }
      });
    });

    describe("when new keyboard is selected", () => {
      let selectedKeyboard: Keyboard;
      beforeEach(() => {
        selectedKeyboard = {
          layout: "gb",
          variant: "dvorak",
        };

        openReactSelectMenu(
          keyboardPage,
          createLabelFromKeyboard(
            defaultProps.currentKeyboard,
            keyboards,
            keyboardVariants
          )
        );

        fireEvent.click(
          getByText(
            createLabelFromKeyboard(
              selectedKeyboard,
              keyboards,
              keyboardVariants
            )
          )
        );
      });

      it("calls setKeyboard with correct keyboard", () => {
        expect(defaultProps.setKeyboard).toHaveBeenCalledWith(selectedKeyboard);
      });
    });
  });
});
