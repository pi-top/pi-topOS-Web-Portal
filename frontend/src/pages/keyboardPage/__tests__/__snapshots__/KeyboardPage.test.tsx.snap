// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyboardPage renders correct image 1`] = `
<img
  alt="keyboard-screen"
  class="image"
  draggable="false"
  src="keyboard-screen.png"
  style="max-height: 100%; max-width: 100%;"
/>
`;

exports[`KeyboardPage renders prompt correctly 1`] = `
<h1
  class="prompt"
>
  QWERTY, 
  <span
    class="green"
  >
    or something else?
  </span>
</h1>
`;
